test/unit/codegen/codegen_ispc.cpp

```diff
@@ -0,0 +1,199 @@
+// Helper for codegen related visitor
+//=============================================================================
+class IspcCodegenTestHelper {
+    /**
+     * Shared pointer to the CodegenIspcVisitor that is used throughout the test
+     */
+    std::shared_ptr<CodegenIspcVisitor> codegen_ispc_visitor;
+
+    /**
+     * Driver to parse the nmodl text
+     */
+    NmodlDriver driver;
+
+    /**
+     * Shared pointer of AST generated by the parsed NMODL text
+     */
+    std::shared_ptr<ast::Program> ast;
+
+    /**
+     * String buffer which saves the output after every call to CodegenIspcVisitor
+     */
+    std::stringbuf strbuf;
+
+  public:
+    /**
+     * \brief Constructs the helper class for the unit tests of CodegenIspcVisitor
+     *
+     * Takes as argument the nmodl text and sets up the CodegenIspcVisitor.
+     *
+     * \param nmodl_text NMODL text to be parsed and printed
+     *
+     */
+    IspcCodegenTestHelper(const std::string& nmodl_text) {
+        ast = driver.parse_string(nmodl_text);
+
+        /// directory where units lib file is located
+        std::string units_dir(NrnUnitsLib::get_path());
+        /// parse units of text
+        UnitsVisitor(units_dir).visit_program(*ast);
+
+        /// construct symbol table
+        SymtabVisitor().visit_program(*ast);
+
+        /// initialize CodegenIspcVisitor
+        std::ostream oss(&strbuf);
+        codegen_ispc_visitor = std::make_shared<CodegenIspcVisitor>(
+            "unit_test", oss, codegen::LayoutType::soa, "double", false);
+        codegen_ispc_visitor->setup(*ast);
+    }
+
+    /**
+     * Get CodegenIspcVisitor to call its functions
+     */
+    std::shared_ptr<CodegenIspcVisitor> get_visitor() {
+        return codegen_ispc_visitor;
+    }
+
+    /**
+     * Returns the existing string and resets the buffer for next call
+     */
+    std::string get_string() {
+        std::string return_string = strbuf.str();
+        strbuf.str("");
+        return return_string;
+    }
+};
+
+std::string print_ispc_nmodl_constants(IspcCodegenTestHelper& ispc_codegen_test_helper) {
+    /// print nmodl constants
+    ispc_codegen_test_helper.get_visitor()->print_nmodl_constants();
+
+    return ispc_codegen_test_helper.get_string();
+}
+
+std::string print_ispc_compute_functions(IspcCodegenTestHelper& ispc_codegen_test_helper) {
+    /// print compute functions
+    ispc_codegen_test_helper.get_visitor()->print_compute_functions();
+
+    return ispc_codegen_test_helper.get_string();
+}
+
+
+SCENARIO("ISPC codegen", "[codegen][ispc]") {
+    GIVEN("Mod file that has multiple double, float and in constants") {
+        std::string nmodl_text = R"(
+            TITLE UnitTest
+            NEURON {
+                SUFFIX unit_test
+                RANGE a, b
+            }
+            ASSIGNED {
+                a
+                b
+            }
+            UNITS {
+                FARADAY = (faraday) (coulomb)
```

Comment:
IIRC this does not cover the whole picture. We should also test:
- R constant
- The possibility to ask for kiloculombs: FARADAY = (faraday) (kilocoulombs)
- Custom value with custom units like: R = 8.3e-3 (kilojoule/degC)
- other constants?
- Do we check with AND without legacy units?

Reply:
All the tests related to the units definition are done in https://github.com/BlueBrain/nmodl/blob/master/test/unit/units/parser.cpp
Here my intention was to just check if the values for the units are printed with the expected form.
I believe the other tests cover all the cases and we don't have to check this also here